tests/integration/exec.go

```diff
@@ -0,0 +1,197 @@
+	}
+	if !filepath.IsAbs(cmd) {
+		cmd, err = filepath.Abs(cmd)
+		if err != nil {
+			return "", nil, err
+		}
+	}
+
+	args := vals
+	// remove single quotes from args, since they can confuse exec
+	for i, arg := range args {
+		args[i] = strings.Trim(arg, "'")
+	}
+
+	return cmd, args, nil
+}
+
+// handleFDs closes all fds except for stdin, stdout, stderr
+// It also redirects stdout to /dev/null
+func handleFDs() error {
+	// close range of fds
+	err := unix.CloseRange(3, ^uint(0), unix.CLOSE_RANGE_UNSHARE)
+	if err != nil {
+		return fmt.Errorf("close range of fds: %w", err)
+	}
+
+	// redirect stdout to /dev/null
+	null, err := os.OpenFile("/dev/null", os.O_RDWR, 0)
+	if err != nil {
+		return fmt.Errorf("open /dev/null: %w", err)
+	}
+	err = syscall.Dup2(int(null.Fd()), unix.Stdout)
+	if err != nil {
+		return fmt.Errorf("dup2: %w", err)
+	}
+	err = syscall.Close(int(null.Fd()))
+	if err != nil {
+		return fmt.Errorf("close: %w", err)
+	}
+
+	return nil
+}
+
+// waitChild waits for a child process to exit
+func waitChild(pid int, timeout time.Duration) (int, error) {
+	var (
+		deadline = time.Now().Add(timeout)
+		wopts    = syscall.WNOHANG | syscall.WUNTRACED | syscall.WCONTINUED
+		wstat    syscall.WaitStatus
+	)
+
+	for {
+		pid, err := syscall.Wait4(pid, &wstat, wopts, nil)
+		if err != nil {
+			return -1, err
+		}
+
+		if pid != 0 { // child process has exited
+			if wstat.Exited() {
+				return pid, nil
+			}
+
+			return -1, fmt.Errorf("child process exited with status: %v", wstat)
+		}
+
+		// check if timeout has occurred
+		if time.Now().After(deadline) {
+			syscall.Kill(pid, syscall.SIGTERM)
+			return -1, fmt.Errorf("timed out waiting for child process to exit")
+		}
+
+		time.Sleep(100 * time.Millisecond) // poll every 100ms
+	}
+}
+
+// forkAndExec forks and execs a command
+// It returns the pid of the child process
+func forkAndExec(cmd string, timeout time.Duration) (int, error) {
+	pid, _, errno := syscall.RawSyscall(syscall.SYS_FORK, 0, 0, 0)
+	if pid == 0 { // child
+		setCPUs(cpuForTests)
+		err := handleFDs()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "handleFDs: %v", err)
+			syscall.Exit(1)
+		}
+
+		cmd, args, err := parseCmd(cmd)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "parseCmd: %v", err)
+			syscall.Exit(1)
+		}
+		fmt.Fprintf(os.Stderr, "\texecuting: %s %v\n", cmd, args)
+
+		// exec
+		err = syscall.Exec(cmd, args, os.Environ())
+		// point of no return, but we need to handle err in case exec fails
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Exec: %v", err)
+			syscall.Exit(1)
+		}
+	}
+
+	if int(pid) == -1 {
+		return -1, syscall.Errno(errno)
+	}
+
+	return waitChild(int(pid), timeout)
+}
+
+// forkAndCallSys forks and calls the syscalls specified in evts, changing the child's comm to newComm
+// It returns the pid of the child process
+func forkAndCallSys(newComm string, evts []trace.Event, timeout time.Duration) (int, error) {
+	pid, _, errno := syscall.RawSyscall(syscall.SYS_FORK, 0, 0, 0)
```

Comment:
I don't think this is as simple as that. Your stack frame might be shared among different runtime threads (even if you're running on a single CPU).

Have you read the implementation of `unix.forkExec()`? The path I'm suggesting you to read, if you haven't, is:

```
forkExec() -> forkAndExecInChild() -> forkAndExecInChild1()
```

You will notice that there are runtime internal functions called:

1. runtime_BeforeFork() (right before vfork() is called)
2. runtime_AfterFork() (right after 
3. runtime_AfterForkInChild()

The clone() syscall is executed by vfork() and not fork(). It means that the parent's page tables won't be copied (likely due to performance, so we would be good in that aspect). The main difference is that the caller is suspended until the child calls execve(). This might make a big difference for runtime (that could be doing a bunch of stuff and preempting the runtime routine calling fork).

There are MANY other things managed by Golang's implementation of forkAndExec(). Many things related to process control group, terminal controlling processing, etc, etc.

It is very likely impossible to have a "forkAndFunction()" approach without the runtime having the capability to do so. A forkAndExec() exists just because runtime allows it to exist (with many mitigations done in the implementation I quoted).

Reply:
> I don't think this is as simple as that. Your stack frame might be shared among different runtime threads (even if you're running on a single CPU).

Even in the case of an ordinary `fork()`? Note that I'm not using `vfork()`.

See below from `man fork 2`:

> The child process and the parent process run in separate memory spaces.  At the time of fork() both memory spaces have  the  same content.   Memory writes, file mappings (mmap(2)), and unmappings (munmap(2)) performed by one of the processes do not affect the other.

Reply:
I meant the "parent" stack frame (shared among multiple runtime threads). This is likely the reason why there are internal (to runtime) functions that need to be called.

![image](https://user-images.githubusercontent.com/7395852/220196650-585ef985-5083-4c04-a248-ea03ffbbb3a8.png)

Reply:
There are runtime locks that will guarantee the stack doesn't grow before an exec takes place, for example, and things like that.

Reply:
I understand that mixing Go runtime with fork and syscalls is not that trivial, especially not using the provided Go API.

My initial tests used a [syscaller.go binary](https://github.com/aquasecurity/tracee/blob/88ac808171c52231ced7bdb82f495ddd4d078f3d/tests/integration/syscaller/syscaller.go) that was detached from the tests runtime and could work nicely. But I noticed that if we wanted to improve the syscalls maps with information brought in from tracee event type it would be hard to maintain (due to code duplication).

Running this PR with runtime.GOMAXPROCS(2), reserving core 0 for tests executions and cores 1..maxcpu for the tests threads (tracee), it seems to work ok.

Instead, what would your proposal be?

Reply:
My suggestion was at the end of comments:

> I think we can have our own version of forkAndExec() setting cpumask and changing command name, etc... but I don't think the forkAndSys would work correctly.

I'm afraid of creating something that "works most of the times" but might introduce races and issues that are only present in the testing environment (because of things we could predict before creating the test environment).

About this:

> But I noticed that if we wanted to improve the syscalls maps with information brought in from tracee event type ...

Could you clarify what would be the info brought from the event type and how you would feed the syscall testing?

Reply:
> I'm afraid of creating something that "works most of the times" but might introduce races and issues that are only present in the testing environment (because of things we could predict before creating the test environment).

I agree with you. So surely my initial detached sycaller.go binary would solve this. Having an external execution, through the ForkAndExec provided by the golang API, would work just fine. If the sycaller broke something related to its runtime, which is not expected anyway, as we will create (chain) the specific calls to be reproducible in each case.

> Could you clarify what would be the info brought from the event type and how you would feed the syscall testing? I ask you this because if its only a syscalls table with a bunch of arguments to be tested, then it might be okay to have something like your syscaller proposal.

```go
var syscallMap = map[events.ID]sysArgs{
	events.Read: {0, 0, 0, 0, 0, 0},
}
```

That is the current implementation, just a raw table with events and args, for simple tests as shown in test table. We can make sure that this simple map can be set with safe events args that won't harm runtime.

Having a foundation working for that, we can add and make use of:

```go
var chainedCallsMap = map[events.ID][]syscall{
	events.Read: {
		events.Open: {pathname, flags, 0, 0, 0, 0},
		events.Read: {fd, buf, count, 0, 0, 0},
		events.Close: {fd, 0, 0, 0, 0, 0},
	},
	// Signatures
	events.AntiDebugging: {
		...,
		events.Ptrace: {whatever, 0, 0, 0, 0, 0},
		...,
	},
}
```

> (which is actually close to having the shell scripts, no ?)

With only shell scripts or ready third-party binaries, we can't have the freedom to specify events in the order and way we want.

> Don't get me wrong, I think its a great idea to have what you are proposing, I just don't think its "safe" enough when it comes to running it over and over and over (and picking a runtime race because of something broke by design).

With a detached syscaller, as I mentioned, we can have a safe runtime.

Regarding the race conditions and the likely go routine leak (test cases loop), I have to dig into it as this behaviour seems to come from the main branch.

Reply:
_An addendum:_ to avoid getting stuck with the processorID test, which doesn't make much sense, we can certainly let the threads run free without blocking them. But we still need to make sure that calls (forks) should only run on the same CPU to avoid unordered events.

Reply:
> With only shell scripts or ready third-party binaries, we can't have the freedom to specify events in the order and way we want.

I understand you want more control of what is being executed and the results it provides (like calling a specific syscall and checking the event it generated before moving on). If that is the case, I agree that having that would be very nice indeed.

Reply:
About running on the same CPU to guarantee ordering, now we have network events and they run from softIRQ context... so that might not be enough for the network event testing... as the runtime might get preempted (if you're thinking about parallelizing the tests and checking results in parallel).

Maybe you will have to stick with a single routine doing event, checking, doing event, checking... but you will find out during implementation.

Reply:
The current proposed solution makes use of an external tool, so I consider this solved.